fetchFromGitHub: add archiveReproducibilityWorkaround option

YorikSar opened the pull request:
This option fetches tarball from GitHub based on tree hash instead of the tag to get unprocessed data from the repo, and then applies a partial implementation of `export-subst` to replace certain format strings, but unlike Git uses reproducible values for them.
This includes an example of using this in a package affected by this non-reproducibility.

Related PR: https://github.com/NixOS/nixpkgs/pull/417859
Related issue: https://github.com/NixOS/nixpkgs/issues/84312
Inspired by discussion: https://discourse.nixos.org/t/fetchfromgithub-and-the-versioneer-fixing-source-reproducibility/66539

### TODO

- [x] move script to a separate package
- [ ] add docs
- [ ] add release notes entry

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

YorikSar committed:
fetchFromGitHub: add archiveReproducibilityWorkaround option

This option fetches tarball from GitHub based on tree hash instead of
the tag to get unprocessed data from the repo, and then applies a
partial implementation of `export-subst` to replace certain format
strings, but unlike Git uses reproducible values for them.

This includes an example of using this in a package affected by this
non-reproducibility.

Related PR: https://github.com/NixOS/nixpkgs/pull/417859
Related issue: https://github.com/NixOS/nixpkgs/issues/84312
Inspired by discussion: https://discourse.nixos.org/t/fetchfromgithub-and-the-versioneer-fixing-source-reproducibility/66539

emilazy commented:
Can we get away with not doing the replacements at all for simplicity? A `git clone` won’t apply them, right? `jq` generating a `sed` script scares me somewhat.

roberth commented on pkgs/build-support/fetchgithub/default.nix:
```diff
@@ -124,6 +143,95 @@ lib.makeOverridable (
+                t: .tree.sha[:7],
+                P: [.parents[] | .sha] | join(" "),
+                p: [.parents[] | .sha[:7]] | join(" "),
+                an: .author.name,
```
Is this the Git name or the GitHub name? The latter would not be pure.

roberth commented on pkgs/build-support/fetchgithub/default.nix:
```diff
@@ -124,6 +143,95 @@ lib.makeOverridable (
+                ae: .author.email,
+                ai: .author.date, # man calls this "ISO 8601-like format", but we just use ISO 8601 provided by GitHub
+                aI: .author.date,
+                cn: .committer.name,
```
I assume the same answer will apply here

roberth commented on pkgs/build-support/fetchgithub/default.nix:
```diff
@@ -21,6 +24,7 @@ lib.makeOverridable (
     sparseCheckout ? [ ],
     githubBase ? "github.com",
     varPrefix ? null,
+    archiveReproducibilityWorkaround ? false,
```
I don't think we should call this a workaround. It is a reimplementation of export-subst, and ideally that would have been something that's exposed by git. Other than that, this is a good solution to the problem, even if the constraints aren't very fun.

```suggestion
    pureExportSubst ? false,
```

roberth commented on pkgs/build-support/fetchgithub/default.nix:
```diff
@@ -30,6 +34,18 @@ lib.makeOverridable (
+      && !fetchSubmodules
+      && (leaveDotGit == null)
+      && !deepClone
+      && !forceFetchGit
```
```suggestion
      && !forceFetchGit  # could be done, but not implemented
```

roberth reviewed (commented):
This looks like a good strategy to solve the export-subst problem.

Could you:
- factor out the processing logic into a separate script package with a test, and/or
- add a fetcher test like we have for fetchgit

https://github.com/NixOS/nixpkgs/blob/1f91f2e538e46cac4398eb16e79a1ee0f698caee/pkgs/test/default.nix#L133

Those use https://nixos.org/manual/nixpkgs/unstable/#tester-invalidateFetcherByDrvHash to generate unique FODs per fetcher implementation "version". Otherwise you end up testing nothing at all after running it once.

Unfortunately, we don't have any for `fetchFromGitHub` yet. Yikes.

YorikSar reviewed (commented):
@roberth, I've addressed all your comments, except I didn't move the script to a separate package yet, will do this later. I'll add a TODO list to the top post.

@emilazy While some packages don't really use this, packages with user interface (CLI, WebUI, etc) might display version string, and it will show the template instead of the actual version. While we could fix this package by package as we currently do, I agree with @roberth's [comment](https://discourse.nixos.org/t/fetchfromgithub-and-the-versioneer-fixing-source-reproducibility/66539/4) that this warrants a more generic fix.

emilazy commented:
> @emilazy While some packages don't really use this, packages with user interface (CLI, WebUI, etc) might display version string, and it will show the template instead of the actual version. While we could fix this package by package as we currently do, I agree with @roberth's [comment](https://discourse.nixos.org/t/fetchfromgithub-and-the-versioneer-fixing-source-reproducibility/66539/4) that this warrants a more generic fix.

But there’s nothing stopping us only implementing the raw fetch inside `fetchFromGitHub` and doing the `export-subst` stuff as a hook outside of the FOD, to avoid having to keep byte‐stability for all eternity, right?

(Even then I’d prefer to wait to see if anything warrants it – we already do a lot of patching of packages for bad version checks and it’s not clear to me this would be a meaningful additional burden – but it wouldn’t have the serious issues of putting more logic in FODs, something we’re trying to get away from as it keeps causing us issues.)

YorikSar committed:
pure-export-subst: init

YorikSar commented:
I've extracted the substitution logic into a separate package and rebased the branch on top of master to resolve a conflict.

Could you please point me to the argument against adding logic to FOD? We seem to run rather complex computations in them anyway (e.g. any `git clone`), so what's one more `sed` going to do?

I'm not against moving this outside of FODs, but there would be a problem of getting data for `export-subst`. In the current state, we're fetching it from the GitHub API, but outside of FOD, we'd have to make users provide all this data somehow, which would force package maintainers to keep this set of parameters up to date somehow.
As for API, I'd prefer to keep it in `fetchFromGitHub` and add these layers of derivations inside of it. I think it would keep API for package maintainers simple.
